File: core/controllers/video_list.py
# coding: utf-8
#
# Copyright 2014 The Oppia Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS-IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Controllers for the videioList page."""
import jinja2

from core.controllers import base
# from core.domain import user_services
# from core.domain import config_domain
from core.controllers import editor
from core.domain import config_domain
from core.domain import dependency_registry
from core.domain import gadget_registry
from core.domain import interaction_registry
from core.domain import rte_component_registry
from core.domain import video_list_service
from core.storage.video_list import gae_models as video_list_models
import feconf


# import utils


class VedioListHandler(base.BaseHandler):
    """处理视频列表"""
    PAGE_NAME_FOR_CSRF = "editor"

    def get(self):
        self.render_json(self.values)


class VideoListPage(base.BaseHandler):
    PAGE_NAME_FOR_CSRF = "editor"
    EDITOR_PAGE_DEPENDENCY_IDS = []

    def get(self):
        if self.username in config_domain.BANNED_USERNAMES.value:
            raise self.UnauthorizedUserException("")
        else:

            dependencies_html, additional_angular_modules = (
                dependency_registry.Registry.get_deps_html_and_angular_modules(
                    self.EDITOR_PAGE_DEPENDENCY_IDS))

            interaction_templates = (
                rte_component_registry.Registry.get_html_for_all_components())

            gadget_types = gadget_registry.Registry.get_all_gadget_types()
            gadget_templates = (
                gadget_registry.Registry.get_gadget_html(gadget_types))

            self.values.update({
                'meta_description': feconf.SPLASH_PAGE_DESCRIPTION,
                'nav_mode': 'video',
                'value_generators_js': jinja2.utils.Markup(
                    editor.get_value_generators_js()),
                'gadget_templates': jinja2.utils.Markup(gadget_templates),
                'interaction_templates': jinja2.utils.Markup(
                    interaction_templates)
            })
            self.render_template(
                'video_list/video_list.html')


class VideoListData(base.BaseHandler):
    PAGE_NAME_FOR_CSRF = "editor"

    """视频数据处理"""

    def get(self, video_id):
        if video_id is not None and video_id != '0':
            video = video_list_service.get_by_id(video_id)
            self.values.update(video.to_dict())
            self.render_json(self.values)
        else:
            urlsafe_start_cursor = self.request.get('cursor')
            lists, new_urlsafe_start_cursor, more = \
                video_list_service.get_all_video(
                    urlsafe_start_cursor=urlsafe_start_cursor)
            self.render_json({
                'results': [m.to_dict() for m in lists],
                'cursor': new_urlsafe_start_cursor,
                'more': more,
            })

    def post(self, video_id):
        name = self.payload.get('name')
        category = self.payload.get('category')
        ids = self.payload.get('ids')
        if video_id is not None and video_id != '0':
            video = video_list_service.get_video_model(video_id)
            video.name = name
            video.category = category
            video.ids = ids
            video.put()
        else:
            video_list_service.create_video(self.user_id, name, category, ids)
        self.render_json(self.values)

    def delete(self, video_id):
        """删除视频"""
        video_list_service.delete_video(video_id)


class VideoCategoryList(base.BaseHandler):
    PAGE_NAME_FOR_CSRF = "editor"
    EDITOR_PAGE_DEPENDENCY_IDS=[]

    def get(self):
        if self.username in config_domain.BANNED_USERNAMES.value:
            raise self.UnauthorizedUserException("用户被禁止访问")
        else:
            dependencies_html, additional_angular_modules = (
                dependency_registry.Registry.get_deps_html_and_angular_modules(
                    self.EDITOR_PAGE_DEPENDENCY_IDS))

            interaction_templates = (
                rte_component_registry.Registry.get_html_for_all_components())

            gadget_types = gadget_registry.Registry.get_all_gadget_types()
            gadget_templates = (
                gadget_registry.Registry.get_gadget_html(gadget_types))

            self.values.update({
                'meta_description': feconf.SPLASH_PAGE_DESCRIPTION,
                'nav_mode': 'video',
                'value_generators_js': jinja2.utils.Markup(
                    editor.get_value_generators_js()),
                'gadget_templates': jinja2.utils.Markup(gadget_templates),
                'interaction_templates': jinja2.utils.Markup(
                    interaction_templates)
            })
            self.render_template(
                'video_list/videoCatelogList.html')


class VideoCategoryData(base.BaseHandler):
    PAGE_NAME_FOR_CSRF = "editor"

    def get(self, category_id):
        urlsafe_start_cursor = self.request.get('cursor')
        lists, new_urlsafe_start_cursor, more = \
            video_list_service.get_all_video_category(
                urlsafe_start_cursor=urlsafe_start_cursor)
        self.render_json({
            'results': [m.to_dict() for m in lists],
            'cursor': new_urlsafe_start_cursor,
            'more': more,
        })

    def post(self, category_id):
        name = self.payload.get('name')
        category = self.payload.get('category')
        ids = self.payload.get('ids')
        if category_id is not None and category_id != '0':
            video = video_list_models.VideoCategory.get(long(category_id))
        else:
            video = video_list_models.VideoCategory()
            video.author_id = self.user_id
        video.name = name
        video.category = category
        video.ids = ids
        video.put()
        self.render_json(self.values)


class VideoView(base.BaseHandler):
    PAGE_NAME_FOR_CSRF = "view"
    EDITOR_PAGE_DEPENDENCY_IDS = []

    def get(self, video_id):
        if self.username in config_domain.BANNED_USERNAMES.value:
            raise self.UnauthorizedUserException("")
        else:

            interaction_templates = (
                rte_component_registry.Registry.get_html_for_all_components())

            gadget_types = gadget_registry.Registry.get_all_gadget_types()
            gadget_templates = (
                gadget_registry.Registry.get_gadget_html(gadget_types))

            self.values.update({
                'meta_description': feconf.SPLASH_PAGE_DESCRIPTION,
                'nav_mode': 'video',
                'value_generators_js': jinja2.utils.Markup(
                    editor.get_value_generators_js()),
                'gadget_templates': jinja2.utils.Markup(gadget_templates),
                'interaction_templates': jinja2.utils.Markup(
                    interaction_templates)
            })
            if video_id is not None and video_id != '0':
                video = video_list_service.get_by_id(video_id)
                self.values.update(video.to_dict())
                self.render_template("video_list/video.html")
            else:
                raise self.PageNotFoundException(None)

File: core/templates/dev/head/video_list/VideoList.js
// Copyright 2015 The Oppia Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the 'License');
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an 'AS-IS' BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/**
 * @fileoverview Translation functions for Oppia.
 *
 * @author milagro.teruel@gmail.com (Milagro Teruel)
 */

// Translations of strings that are loaded in the front page. They are listed

oppia.controller('VideoList', ['$scope', '$modal', '$mdDialog', '$rootScope', '$window',
  'oppiaDatetimeFormatter', 'alertsService', 'FATAL_ERROR_CODES',
  'videoListService', '$sce',
  function($scope, $modal, $mdDialog, $rootScope, $window,
           oppiaDatetimeFormatter, alertsService,
           FATAL_ERROR_CODES, videoListService, $sce) {
    var _cursor = null;
    $scope.endOfPageIsReached = false;
    $scope.loadData = function() {
      $rootScope.loadingMessage = '加载中';
      videoListService.getVideoList().then(function(response) {
        $scope.videos = response.data.results;
        _cursor = response.data.cursor;
        $scope.endOfPageIsReached = !response.data.more;
        $scope.currentUserIsAdmin = response.data.is_admin;
        $scope.currentUserName = response.data.username;
        $scope.category_objective = $sce.trustAsHtml(response.data.category_objective);
        $rootScope.loadingMessage = '';
      });
    };
    $scope.getLocaleAbbreviatedDatetimeString = function(millisSinceEpoch) {
      return oppiaDatetimeFormatter.getLocaleAbbreviatedDatetimeString(
        millisSinceEpoch);
    };
    $scope.loadData();
    $scope.showMore = function() {
      if (!$rootScope.loadingMessage) {
        videoListService.getVideoMore(_cursor).then(function(response) {
          $scope.videos = $scope.videos.concat(response.data.results);
          _cursor = response.data.cursor;
          $scope.endOfPageIsReached = !response.data.more;
          $scope.currentUserIsAdmin = response.data.is_admin;
          $scope.currentUserName = response.data.username;
          $scope.category_objective = $sce.trustAsHtml(response.data.category_objective);
          $rootScope.loadingMessage = '';
        });
      }
    };
    $scope.showEditModal = function(objid) {
      videoListService.getVideo(objid).then(function(response) {
        $modal.open({
          templateUrl: 'modals/editorVideoCreate',
          backdrop: true,
          resolve: {},
          controller: ['$scope', '$modalInstance', 'CATEGORY_LIST',
            'ALL_CATEGORIES_ZH_MAP',
            function($scope, $modalInstance, CATEGORY_LIST,
                     ALL_CATEGORIES_ZH_MAP) {
              $scope.schema = {
                type: 'html'
              };
              $scope.name = response.data.name;
              $scope.ids = response.data.ids;
              $scope.id = response.data.id;
              $scope.create = function(name, ids) {
                $modalInstance.close({
                  id: $scope.id,
                  name: name,
                  ids: ids,
                  category: categoryId
                });
              };
              $scope.cancel = function() {
                $modalInstance.dismiss('cancel');
              };
            }]
        }).result.then(function(result) {
          $rootScope.loadingMessage = '正在保存';
          videoListService.saveEditVideo(result.id, result.name,
            result.ids,
            result.category).then(function() {
            $rootScope.loadingMessage = '';
            alertsService.addSuccessMessage('保存成功');
            $scope.loadData();
          }, function() {
            $rootScope.loadingMessage = '';
            alertsService.addWarning('保存失败.');
          });
        });
      }, function() {
        $rootScope.loadingMessage = '';
        alertsService.addWarning('信息不存在.');
      });
    };
    $scope.showCreateModal = function() {
      $modal.open({
        templateUrl: 'modals/editorVideoCreate',
        backdrop: true,
        resolve: {},
        controller: ['$scope', '$modalInstance', 'CATEGORY_LIST',
          'ALL_CATEGORIES_ZH_MAP',
          function($scope, $modalInstance, CATEGORY_LIST,
                   ALL_CATEGORIES_ZH_MAP) {
            $scope.schema = {
              type: 'html'
            };
            $scope.create = function(name, ids) {
              $modalInstance.close({
                name: name,
                ids: ids,
                category: categoryId
              });
            };
            $scope.cancel = function() {
              $modalInstance.dismiss('cancel');
            };
          }]
      }).result.then(function(result) {
        $rootScope.loadingMessage = '正在保存';
        videoListService.createVideo(result.name,
          result.ids,
          result.category).then(function() {
          $rootScope.loadingMessage = '';
          alertsService.addSuccessMessage('保存成功');
          $scope.loadData();
        }, function() {
          $rootScope.loadingMessage = '';
          alertsService.addWarning('保存失败.');
        });
      });
    };
    $scope.deleteData = function(objid, ev) {
      ev.stopPropagation();
      var confirm = $mdDialog.confirm({
        template: [
          '<md-dialog aria-label="<[dialog.label]>">',
          '<md-content>',
          '<h2><[ dialog.title ]></h2>',
          '<p><[ dialog.content ]></p>',
          '</md-content>',
          '<div class="md-actions">',
          '<md-button ng-if="dialog.$type == \'confirm\'" ng-click="dialog.abort()">',
          '<[ dialog.cancel ]>',
          '</md-button>',
          '<md-button ng-click="dialog.hide()" class="md-primary">',
          '<[ dialog.ok ]>',
          '</md-button>',
          '</div>',
          '</md-dialog>'
        ].join('')
      })
        .title('确认删除?')
        .content('您确定要删除数据吗，删除后不能恢复！')
        .ariaLabel('Lucky day')
        .ok('确定')
        .cancel('取消').targetEvent(ev);
      $mdDialog.show(confirm).then(function() {
        videoListService.deleteData(objid).then(function(response) {
          var data = response.data;
          alertsService.addSuccessMessage('删除成功!');
          $scope.loadData();
        })
      }, function() {
      });
    };
  }
]);

oppia.controller('VideoCategoryList', ['$scope', '$modal', '$mdDialog',
  '$rootScope', '$window',
  'oppiaDatetimeFormatter',
  'alertsService', 'FATAL_ERROR_CODES',
  'VideoCategoryService',
  function($scope, $modal, $mdDialog, $rootScope, $window,
           oppiaDatetimeFormatter, alertsService,
           FATAL_ERROR_CODES, VideoCategoryService) {
    var _cursor = null;
    $scope.endOfPageIsReached = false;
    $scope.loadData = function() {
      $rootScope.loadingMessage = '加载中';
      VideoCategoryService.getList().then(function(response) {
        $scope.videos = response.data.results;
        $scope.endOfPageIsReached = !response.data.more;
        _cursor = response.data.cursor;
        $scope.currentUserName = response.data.username;
        $scope.currentUserIsAdmin = response.data.is_admin;
        $rootScope.loadingMessage = '';
      });
    };
    $scope.getLocaleAbbreviatedDatetimeString = function(millisSinceEpoch) {
      return oppiaDatetimeFormatter.getLocaleAbbreviatedDatetimeString(
        millisSinceEpoch);
    };
    $scope.loadData();
    $scope.showMore = function() {
      if (!$rootScope.loadingMessage) {
        VideoCategoryService.getListMore(_cursor).then(function(response) {
          $scope.videos = $scope.videos.concat(response.data.results);
          $scope.endOfPageIsReached = !response.data.more;
          _cursor = response.data.cursor;
        })
      }
    };
    $scope.showCreateModal = function() {
      $modal.open({
        templateUrl: 'modals/editorVideoCreate',
        backdrop: true,
        resolve: {},
        controller: ['$scope', '$modalInstance', 'CATEGORY_LIST',
          'ALL_CATEGORIES_ZH_MAP',
          function($scope, $modalInstance, CATEGORY_LIST,
                   ALL_CATEGORIES_ZH_MAP) {
            $scope.schema = {
              type: 'html',
              ui_config: {
                'hide_complex_extensions': true
              }
            };
            $scope.CATEGORY_LIST_FOR_SELECT2 = [];

            for (var i = 0; i < CATEGORY_LIST.length; i++) {
              $scope.CATEGORY_LIST_FOR_SELECT2.push({
                id: CATEGORY_LIST[i],
                text: ALL_CATEGORIES_ZH_MAP[CATEGORY_LIST[i]]
              });
            }
            $scope.create = function(name, pictureName,
                                     category, objective) {
              $modalInstance.close({
                name: name,
                picture_name: pictureName,
                category: category,
                objective: objective
              });
            };
            $scope.cancel = function() {
              $modalInstance.dismiss('cancel');
            };
          }]
      }).result.then(function(result) {
        $rootScope.loadingMessage = '正在保存';
        VideoCategoryService.create(result.name,
          result.picture_name,
          result.category, result.objective).then(function() {
          $rootScope.loadingMessage = '';
          alertsService.addSuccessMessage('保存成功');
          $scope.loadData();
        }, function() {
          $rootScope.loadingMessage = '';
          alertsService.addWarning('保存失败.');
        });
      });
    };
    $scope.showEditModal = function(objid) {

      $rootScope.loadingMessage = '正在加载';
      VideoCategoryService.get(objid).then(function(response) {
        $rootScope.loadingMessage = '';
        $modal.open({
          templateUrl: 'modals/editorVideoCreate',
          backdrop: true,
          resolve: {},
          controller: ['$scope', '$modalInstance', 'CATEGORY_LIST',
            'ALL_CATEGORIES_ZH_MAP',
            function($scope, $modalInstance, CATEGORY_LIST,
                     ALL_CATEGORIES_ZH_MAP) {
              $scope.schema = {
                type: 'html',
                ui_config: {
                  'hide_complex_extensions': true
                }
              };
              $scope.CATEGORY_LIST_FOR_SELECT2 = [];

              for (var i = 0; i < CATEGORY_LIST.length; i++) {
                $scope.CATEGORY_LIST_FOR_SELECT2.push({
                  id: CATEGORY_LIST[i],
                  text: ALL_CATEGORIES_ZH_MAP[CATEGORY_LIST[i]]
                });
              }
              var data = response.data;
              $scope.name = data.name;
              $scope.id = data.id;
              $scope.picture_name = data.picture_name;
              $scope.created_on = data.created_on;
              $scope.category = data.category;
              $scope.objective = data.objective;
              $scope.a = {};
              $scope.a.objective = data.objective;
              $scope.create = function(name, pictureName,
                                       category, objective) {
                $modalInstance.close({
                  name: name,
                  picture_name: pictureName,
                  category: category,
                  objective: objective,
                  id: $scope.id
                });
              };
              $scope.cancel = function() {
                $modalInstance.dismiss('cancel');
              };
            }]
        }).result.then(function(result) {
          $rootScope.loadingMessage = '正在保存';
          VideoCategoryService.save(result.id, result.name,
            result.picture_name,
            result.category, result.objective).then(function() {
            $rootScope.loadingMessage = '';
            $scope.loadData();
            alertsService.addSuccessMessage('保存成功');
          }, function() {
            $rootScope.loadingMessage = '';
            alertsService.addWarning('保存失败.');
          });
        });
      }, function() {
        alertsService.addWarning('获取数据失败.');
      });
    };
    $scope.deleteData = function(objid, ev) {
      var confirm = $mdDialog.confirm({
        template: [
          '<md-dialog aria-label="<[dialog.label]>">',
          '<md-content>',
          '<h2><[ dialog.title ]></h2>',
          '<p><[ dialog.content ]></p>',
          '</md-content>',
          '<div class="md-actions">',
          '<md-button ng-if="dialog.$type == \'confirm\'" ng-click="dialog.abort()">',
          '<[ dialog.cancel ]>',
          '</md-button>',
          '<md-button ng-click="dialog.hide()" class="md-primary">',
          '<[ dialog.ok ]>',
          '</md-button>',
          '</div>',
          '</md-dialog>'
        ].join('')
      })
        .title('确认删除?')
        .content('您确定要删除数据吗，删除后不能恢复！')
        .ariaLabel('Lucky day')
        .ok('确定')
        .cancel('取消').targetEvent(ev);
      $mdDialog.show(confirm).then(function() {
        VideoCategoryService.deleteData(objid).then(function(response) {
          var data = response.data;
          alertsService.addSuccessMessage('删除成功!');
          $scope.loadData();
        })
      }, function() {
      });
    };
  }
]);
